### DISCUSSION TITLE
Is there a way to detect that newer performance optimizations are enabled?

### DISCUSSION THREAD

**Discussion by eamodio** [git-for-windows/git#4017] [Q&A]
We are working on a new [Commit Graph in GitLens](https://github.com/gitkraken/vscode-gitlens/discussions/2158) and some newer performance optimizations like the commit-graph and file watching can have a very large impact on performance for some repositories. For example, writing a commit-graph for the linux repo on my machine cut the time to open our Graph from ~8s to ~1 (with the same reduction on paging in new data too).

As far as I can tell, there doesn't seem to be a git command to tell me if there is a `commit-graph` or tell if `maintenance` is setup. For the FSMonitor I can read the `[core.fsmonitor](https://git-scm.com/docs/git-config#Documentation/git-config.txt-corefsmonitor)` setting at least.

Should I be looking for the existence of the `commit-graph` file itself, but then how could I tell how it was generated, e.g. with `--reachable` or `--changed-paths` etc.

Any guidance would be greatly appreciated!

**Comment by derrickstolee** [comment]
To determine if these things are available, Git itself uses its config and parsing code to detect first "Is this feature enabled?" and second "Is this feature available?"

I mention this in that the `commit-graph` is enabled by default, but might not actually be written locally. Git's checks are "is `core.commitGraph` enabled?" and then "does the `commit-graph` file exist?" You can consider looking into the `$GIT_DIR/objects/info/` for a `commit-graph` file or a `commit-graphs` directory. Looking into Git at that level is fragile as Git can change things (usually by adding new formats, but not by removing those as possibilities).

**Reply by derrickstolee** [reply]
Oh, and you can't tell that it was built by `--reachable` or not, but the `--changed-paths` option adds extra data to the `commit-graph` file. I don't think there is any way to find out about that kind of internal information any other way.

**Reply by eamodio** [reply]
Thanks for the info. Do you know of any plans to track that metadata as to how it was created?
